KubeArmor/BPF/system_monitor.c

```diff
@@ -1296,4 +1302,140 @@ int kretprobe__listen(struct pt_regs *ctx)
+
+    init_context(&context);
+    context.argnum = get_arg_num(types);
+    
+    switch (conn.skc_family)
+    {
+    case AF_INET:   
+        
+        sockv4.sin_family = conn.skc_family;
+        sockv4.sin_port = conn.skc_dport;
+        sockv4.sin_addr.s_addr = conn.skc_daddr;
+        args.args[1] = (unsigned long) &sockv4;
+        context.event_id = _TCP_CONNECT;
+        break;
+    case AF_INET6:
+        
+        sockv6.sin6_family = conn.skc_family;
+        sockv6.sin6_port = conn.skc_dport;
+        bpf_probe_read(&sockv6.sin6_addr.in6_u.u6_addr16, sizeof(sockv6.sin6_addr.in6_u.u6_addr16), conn.skc_v6_daddr.in6_u.u6_addr16);
+        args.args[1] = (unsigned long) &sockv6;
+        context.event_id = _TCP_CONNECT_v6;
+        break;
+
+    default:
+        return 0;
+        break;
+    }
+
+    args.args[0] = (unsigned long) conn.skc_prot->name;
+    set_buffer_offset(DATA_BUF_TYPE, sizeof(sys_context_t));
+    bufs_t *bufs_p = get_buffer(DATA_BUF_TYPE);
+    if (bufs_p == NULL)
+        return 0;
+    save_context_to_buffer(bufs_p, (void*)&context);
+    save_args_to_buffer(types, &args);
+    events_perf_submit(ctx);
+
+    return 0;
+}
+
+
+SEC("kretprobe/__x64_sys_inet_csk_accept")
+int kretprobe__inet_csk_accept(struct pt_regs *ctx){
+    struct sock *newsk = (struct sock *)PT_REGS_RC(ctx);    
+    if (newsk == NULL)
+        return 0;
+    
+// Code from https://github.com/iovisor/bcc/blob/master/tools/tcpaccept.py with adaptations
+    u16 protocol = 1;
+    int gso_max_segs_offset = offsetof(struct sock, sk_gso_max_segs);
+    int sk_lingertime_offset = offsetof(struct sock, sk_lingertime);
+
+if (sk_lingertime_offset - gso_max_segs_offset == 2)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 6, 0)
+        protocol = READ_KERN(newsk->sk_protocol);
+#else
+        protocol = newsk->sk_protocol;
+#endif
+    else if (sk_lingertime_offset - gso_max_segs_offset == 4)
+        // 4.10+ with little endian
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+        protocol = READ_KERN(*(u8 *)((u64)&newsk->sk_gso_max_segs - 3));
+    else
+        // pre-4.10 with little endian
+        protocol = READ_KERN(*(u8 *)((u64)&newsk->sk_wmem_queued - 3));
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+        // 4.10+ with big endian
+        protocol = READ_KERN(*(u8 *)((u64)&newsk->sk_gso_max_segs - 1));
+    else
+        // pre-4.10 with big endian
+        protocol = READ_KERN(*(u8 *)((u64)&newsk->sk_wmem_queued - 1));
+#else
+# error "Fix your compiler's __BYTE_ORDER__?!"
+#endif   
+  
+    if (protocol != IPPROTO_TCP)
+        return 0;
+    
+
+    struct sock_common conn = READ_KERN(newsk->__sk_common);
+    struct sockaddr_in sockv4;
+    struct sockaddr_in6 sockv6;
+    sys_context_t context = {};
+    args_t args = {};
+    u64 types = ARG_TYPE0(STR_T)|ARG_TYPE1(SOCKADDR_T);
+    init_context(&context);
+    context.argnum = get_arg_num(types);
+
+    switch (conn.skc_family)
```

Comment:
This piece of code seems to be common between connect/accept. Can we refactor it?

Reply:
Done